apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -111,39 +116,40 @@ internal_resolve(Protocol, DNSMessage) ->
     Response = erldns_handler:do_handle(DNSMessage, ?LOCALHOST),
     encode_message(Protocol, Response).
 
--spec(resolve(protocol(), [upstream()], binary(), reply_fun()) -> ok).
-resolve(Protocol, Upstreams, Request, Fun) ->
+-spec(internal_resolve(protocol(), [upstream()], binary(), reply_fun(), binary())-> ok).
+internal_resolve(Protocol, Upstreams, Request, Fun, Zone) ->
     Workers =
         lists:map(fun (Upstream) ->
             Pid = start_worker(Protocol, Upstream, Request),
             MonRef = monitor(process, Pid),
             {Upstream, Pid, MonRef}
         end, Upstreams),
-    resolve_loop(Workers, Fun).
+    resolve_loop(Workers, Fun, Zone).
 
--spec(resolve_loop([{upstream(), pid(), reference()}], reply_fun()) -> ok).
-resolve_loop([], _Fun) ->
+-spec(resolve_loop([{upstream(), pid(), reference()}], reply_fun(), binary()) -> ok).
+resolve_loop([], _Fun, _Zone) ->
     ok;
-resolve_loop(Workers, Fun) ->
+resolve_loop(Workers, Fun, Zone) ->
```

Comment:
- Improve the way this loop works as to deal with timeouts from independent workers

Reply:
Look at PR #47